modules/twinkleblock.js

```diff
@@ -0,0 +1,879 @@
+          checked: true
+        },
+        {
+          label: 'Add block template to user talk page',
+          value: 'template',
+          tooltip: 'If the blocking admin forgot to issue a block template, or you have just blocked the user without templating them, you can use this to issue the appropriate template.',
+          checked: true
+        }
+      ]
+    });
+
+  form.append({ type: 'field', label: 'Preset', name: 'field_preset' });
+  form.append({ type: 'field', label: 'Template options', name: 'field_template_options' });
+  form.append({ type: 'field', label: 'Block options', name: 'field_block_options' });
+
+  form.append( { type:'submit' } );
+
+  result = form.render();
+  Window.setContent( result );
+  Window.display();
+
+  Twinkle.block.fetchUserInfo(function() {
+    // init the controls after user and block info have been fetched
+    var evt = document.createEvent( "Event" );
+    evt.initEvent( 'change', true, true );
+    result.actiontype[0].dispatchEvent( evt );
+  });
+};
+
+Twinkle.block.fetchUserInfo = function twinkleblockFetchUserInfo(fn) {
+
+  api.get({
+    format: 'json',
+    action: 'query',
+    list: 'blocks|users',
+    bkusers: mw.config.get('wgRelevantUserName'),
+    ususers: mw.config.get('wgRelevantUserName')
+  })
+  .then(function(data){
+    var blockinfo = data.query.blocks[0],
+        userinfo = data.query.users[0];
+
+    Twinkle.block.isRegistered = !!userinfo.userid;
+    relevantUserName = Twinkle.block.isRegistered ? "User:" + relevantUser : relevantUser;
+
+    if (blockinfo) Twinkle.block.currentBlockInfo = blockinfo;
+
+    if (typeof fn === 'function') return fn();
+  }, function(msg) {
+    Morebits.status.init($('div[name="currentblock"] span').last()[0]);
+    Morebits.status.warn('Error fetching user info', msg);
+  });
+};
+
+Twinkle.block.callback.changeAction = function twinkleblockCallbackChangeAction(e) {
+  var field_preset, field_template_options, field_block_options, $form = $(e.target.form);
+
+  if ($form.find('[name=actiontype][value=block]').is(':checked')) {
+    field_preset = new Morebits.quickForm.element({ type: 'field', label: 'Preset', name: 'field_preset' });
+    field_preset.append({
+        type: 'select',
+        name: 'category',
+        label: 'Choose a preset:',
+        event: Twinkle.block.callback.change_preset,
+        list: [{
+          label: '',
+          value: '',
+          type: 'option'
+        }].concat(Twinkle.block.callback.filtered_block_groups())
+      });
+
+    field_block_options = new Morebits.quickForm.element({ type: 'field', label: 'Block options', name: 'field_block_options' });
+    field_block_options.append({ type: 'div', name: 'currentblock', label: ' ' });
+    field_block_options.append({
+        type: 'select',
+        name: 'expiry_preset',
+        label: 'Expiry:',
+        event: Twinkle.block.callback.change_expiry,
+        list: [
+          { label: 'custom', value: 'custom', selected: true },
```

Comment:
I should also note that the intention with protect (and other sysop modules, like CSD) was to *improve* on the built-in functionality, not simply to replicate it. I don't know if it succeeds in that goal, but that was the idea. So don't feel obliged to precisely replicate core functionality into Twinkle.